packages/core/contracts/financial-templates/optimistic-distributor/OptimisticDistributor.sol

```diff
@@ -0,0 +1,464 @@
+        optimisticOracle.proposePriceFor(
+            msg.sender,
+            address(this),
+            reward.priceIdentifier,
+            timestamp,
+            ancillaryData,
+            int256(1e18)
+        );
+
+        // Store and log proposed distribution.
+        proposals[proposalId] = Proposal({
+            rewardIndex: rewardIndex,
+            timestamp: timestamp,
+            merkleRoot: merkleRoot,
+            ipfsHash: ipfsHash
+        });
+        emit ProposalCreated(
+            reward.sponsor,
+            reward.rewardToken,
+            rewardIndex,
+            timestamp,
+            reward.maximumRewardAmount,
+            proposalId,
+            merkleRoot,
+            ipfsHash
+        );
+    }
+
+    /**
+     * @notice Allows any caller to execute distribution that has been validated by the Optimistic Oracle.
+     * @param proposalId Hash for identifying existing rewards distribution proposal.
+     * @dev Calling this for unresolved proposals will revert.
+     */
+    function executeDistribution(bytes32 proposalId) external nonReentrant() {
+        // All valid proposals should have non-zero proposal timestamp.
+        Proposal memory proposal = proposals[proposalId];
+        require(proposal.timestamp != 0, "Invalid proposalId");
+
+        // Only one validated proposal per reward can be executed for distribution.
+        Reward memory reward = rewards[proposal.rewardIndex];
+        require(reward.distributionProposed != DistributionProposed.Accepted, "Reward already distributed");
+
+        // Append reward index to ancillary data.
+        bytes memory ancillaryData = _appendRewardIndex(proposal.rewardIndex, reward.customAncillaryData);
+
+        // Get resolved price. Reverts if the request is not settled or settleable.
+        int256 resolvedPrice =
+            optimisticOracle.settleAndGetPrice(reward.priceIdentifier, proposal.timestamp, ancillaryData);
+
+        // Transfer rewards to MerkleDistributor for accepted proposal and flag distributionProposed Accepted.
+        if (resolvedPrice == 1e18) {
+            rewards[proposal.rewardIndex].distributionProposed = DistributionProposed.Accepted;
+
+            reward.rewardToken.safeApprove(address(merkleDistributor), reward.maximumRewardAmount);
+            merkleDistributor.setWindow(
+                reward.maximumRewardAmount,
+                address(reward.rewardToken),
+                proposal.merkleRoot,
+                proposal.ipfsHash
+            );
+            emit RewardDistributed(
+                reward.sponsor,
+                reward.rewardToken,
+                proposal.rewardIndex,
+                reward.maximumRewardAmount,
+                proposalId,
+                proposal.merkleRoot,
+                proposal.ipfsHash
+            );
+        } else emit ProposalRejected(proposal.rewardIndex, proposalId);
```

Comment:
Repeated calls to `executeDistribution` would continue emitting `ProposalRejected` for rejected proposals without settlement to proposer/disputer.

Alternative would be to simplify this contract and just `require(resolvedPrice == 1e18)` without emitting `ProposalRejected`, but then the disputer would need to manually claim settlement on OO. Which approach is better?

Reply:
I think what you have here makes sense due to the reason you described. add a comment to say why you did. itthis way and then it's good.

Reply:
added in a83eb842eeddb76aaa96718b74058547a424e778